pytorch_lightning/trainer/connectors/logger_connector/result.py

```diff
@@ -505,3 +531,54 @@ def __getstate__(self) -> dict:
         if minimize is not None:
             d['_minimize'] = minimize.detach()
         return d
+
+    def state_dict(self):
+
+        def to_state_dict(
+            item: Union[ResultMetric, ResultMetricCollection]
+        ) -> Union[_ResultMetricSerializationHelper, _ResultMetricCollectionSerializationHelper]:
+            if isinstance(item, ResultMetricCollection):
+                return _ResultMetricCollectionSerializationHelper(
+                    apply_to_collection(item, ResultMetric, to_state_dict), metadata=item.meta
+                )
+            state = item.__getstate__()
+            state["meta"].sync.fn = None
+            return _ResultMetricSerializationHelper(**item.__getstate__())
```

Comment:
is there no way to avoid pickling one of our objects?
If  one day the implementation or even just the name or import changes, it will not be possible to load the checkpoints, which has given us serious headaches already with callbacks, etc.

Reply:
Great question ! 
@awaelchli While applying `apply_to_collection`, we needed a placeholder to differentiate from pure dict.
I couldn't find a better idea. Do you have one ?

Right now, `_ResultMetricSerializationHelper` should an blank dictionarry and the name shouldn't change.

```py
class _ResultMetricSerializationHelper(dict):    
       """    Since ``ResultCollection`` can hold ``ResultMetric`` values or dictionaries of them, we need    a class to       differentiate between the cases after converting to state dict when saving its state.    """
```

Reply:
@carmocca Any thoughts ?